edgedb.go

```diff
@@ -74,36 +81,98 @@ func connectOne(ctx context.Context, cfg *connConfig, conn *baseConn) error {
 		err error
 	)
 
+	conn.writer = buff.NewWriter()
+
 	for _, addr := range cfg.addrs { // nolint:gocritic
-		// todo do error values need to be checked?
 		conn.conn, err = d.DialContext(ctx, addr.network, addr.address)
 		if err != nil {
-			log.Printf("while attempting connection %+v: %+v", addr, err)
+			err = wrapAll(err)
+			continue
+		}
+
+		toBeDeserialized := make(chan *soc.Data, 2)
+		r := buff.NewReader(toBeDeserialized)
+		go soc.Read(conn.conn, soc.NewMemPool(4, 256*1024), toBeDeserialized)
+
+		err = conn.setDeadline(ctx)
+		if err != nil {
+			err = wrapAll(err, conn.conn.Close())
 			continue
 		}
 
-		err = conn.connect(ctx, cfg)
+		err = conn.connect(r, cfg)
 		if err != nil {
-			_ = conn.conn.Close()
-			log.Printf("while attempting connection %+v: %+v", addr, err)
+			err = wrapAll(err, conn.conn.Close())
 			continue
 		}
 
+		conn.acquireReaderSignal = make(chan struct{}, 1)
+		conn.readerChan = make(chan *buff.Reader, 1)
+		conn.releaseReader(r)
 		return nil
 	}
 
 	conn.conn = nil
 	return err
 }
 
+func (c *baseConn) setDeadline(ctx context.Context) error {
+	deadline, _ := ctx.Deadline()
+	if e := c.conn.SetDeadline(deadline); e != nil {
+		return fmt.Errorf("%v", e)
+	}
+
+	return nil
+}
+
+func (c *baseConn) acquireReader(ctx context.Context) (*buff.Reader, error) {
+	c.acquireReaderSignal <- struct{}{}
+
+	select {
+	case r := <-c.readerChan:
+		if r.Err != nil {
+			return nil, r.Err
+		}
+
+		return r, nil
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+func (c *baseConn) releaseReader(r *buff.Reader) {
+	go func() {
+		for r.Next(c.acquireReaderSignal) {
+			if e := c.fallThrough(r); e != nil {
+				// todo what is the right thing to do here?
+				panic(e)
+			}
+		}
+
+		c.readerChan <- r
+	}()
+}
+
 // Close the db connection
 func (c *baseConn) close() error {
-	return wrapAll(c.terminate(), c.conn.Close())
+	_, err := c.acquireReader(context.Background())
+	return wrapAll(err, c.terminate(), c.conn.Close())
 }
 
 // Execute an EdgeQL command (or commands).
-func (c *baseConn) Execute(ctx context.Context, cmd string) (err error) {
-	return c.scriptFlow(ctx, c.conn, cmd)
+func (c *baseConn) Execute(ctx context.Context, cmd string) error {
+	r, err := c.acquireReader(ctx)
+	if err != nil {
+		return err
+	}
```

Comment:
shouldn't you `defer c.releaseReader(r)` right here?

Reply:
No. The reader should not be released if the connection failed because there is nothing pushing data from the socket to the reader. What should be here but isn't (I over looked this) is a check to see if the socket read produced any errors. if there was an error the socket should be closed or if recoverable the `soc.Read` go routine should be restarted. 

If something panics in `scriptFlow` the socket should be closed and the `baseConn` object should be unusable. Panics should never happen, they indicate a flaw in the driver. (currently this may not be true because https://github.com/edgedb/edgedb-go/issues/24 is not closed yet).

Reply:
I don't understand. You implement literally what I suggested in all other methods like `QueryOne`, what makes `Execute` different?

Reply:
They are also wrong. I over looked the error handling in this PR. I'll update it and have you review again.

Reply:
Updated. I'm not 100% I have it right but I think its better than it was.